source/stream.c

```diff
@@ -0,0 +1,359 @@
+ */
+
+struct aws_input_stream_byte_cursor_impl {
+    struct aws_byte_cursor original_cursor;
+    struct aws_byte_cursor current_cursor;
+};
+
+/*
+ * This is an ugly function that, in the absence of better guidance, is designed to handle all possible combinations of
+ *  aws_off_t (int32_t, int64_t) x all possible combinations of size_t (uint32_t, uint64_t).  Whether the anomalous
+ * combination of int64_t vs. uint32_t is even possible on any real platform is unknown.  If size_t ever exceeds 64 bits
+ * this function will fail badly.
+ *
+ *  Safety and invariant assumptions are sprinkled via comments.  The overall strategy is to cast up to 64 bits and
+ * perform all arithmetic there, being careful with signed vs. unsigned to prevent bad operations.
+ *
+ *  Assumption #1: aws_off_t resolves to a signed integer 64 bits or smaller
+ *  Assumption #2: size_t resolves to an unsigned integer 64 bits or smaller
+ */
+static int s_aws_input_stream_byte_cursor_seek(
+    struct aws_input_stream *stream,
+    aws_off_t offset,
+    enum aws_stream_seek_basis basis) {
+    struct aws_input_stream_byte_cursor_impl *impl = stream->impl;
+
+    uint64_t final_offset = 0;
+    int64_t checked_offset = offset; /* safe by assumption 1 */
+
+    switch (basis) {
+        case AWS_SSB_BEGIN:
+            /*
+             * (uint64_t)checked_offset -- safe by virtue of the earlier is-negative check + Assumption 1
+             * (uint64_t)impl->original_cursor.len -- safe via assumption 2
+             */
+            if (checked_offset < 0 || (uint64_t)checked_offset > (uint64_t)impl->original_cursor.len) {
+                return aws_raise_error(AWS_IO_STREAM_INVALID_SEEK_POSITION);
+            }
+
+            /* safe because negative offsets were turned into an error */
+            final_offset = (uint64_t)checked_offset;
+            break;
+
+        case AWS_SSB_END:
+            /*
+             * -checked_offset -- safe as long checked_offset is not INT64_MIN which was previously checked
+             * (uint64_t)(-checked_offset) -- safe because (-checked_offset) is positive (and < INT64_MAX < UINT64_MAX)
+             */
+            if (checked_offset > 0 || checked_offset == INT64_MIN ||
+                (uint64_t)(-checked_offset) > (uint64_t)impl->original_cursor.len) {
+                return aws_raise_error(AWS_IO_STREAM_INVALID_SEEK_POSITION);
+            }
+
+            /* cases that would make this unsafe became errors with previous conditional */
+            final_offset = (uint64_t)impl->original_cursor.len - (uint64_t)(-checked_offset);
+            break;
+    }
+
+    /* true because we already validated against (impl->original_cursor.len) which is <= SIZE_MAX */
+    assert(final_offset <= SIZE_MAX);
+
+    impl->current_cursor = impl->original_cursor;
+
+    /* (size_t) final_offset - safe via previous assert's reasoning */
+    aws_byte_cursor_advance(&impl->current_cursor, (size_t)final_offset);
```

Comment:
not-fun fact: aws_byte_cursor_advance() fails if distance > SIZE_MAX/2

Reply:
advance-be-gone